swap/simulations_test.go

```diff
@@ -0,0 +1,695 @@
+			break
+		}
+		// don't overheat the CPU...
+		time.Sleep(5 * time.Millisecond)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+
+	maxCheques := 42
+
+	p2Peer := ts1.peers[p2]
+	p1Peer := ts2.peers[p1]
+
+	for i := 0; i < maxCheques; i++ {
+		if i%2 == 0 {
+			p2Peer.Send(ctx, &testMsgBigPrice{})
+			err := waitForChequeProcessed(ts2)
+			if err != nil {
+				t.Fatal(err)
+			}
+		} else {
+			p1Peer.Send(ctx, &testMsgBigPrice{})
+			err := waitForChequeProcessed(ts1)
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+	}
+
+	ch1, err := ts2.swap.loadLastReceivedCheque(p1)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ch2, err := ts1.swap.loadLastReceivedCheque(p2)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expected := uint64(maxCheques / 2 * (DefaultPaymentThreshold + 1))
+	if ch1.CumulativePayout != expected {
+		t.Fatalf("expected cumulative payout to be %d, but is %d", expected, ch1.CumulativePayout)
+	}
+	if ch2.CumulativePayout != expected {
+		t.Fatalf("expected cumulative payout to be %d, but is %d", expected, ch2.CumulativePayout)
+	}
+
+	log.Info("Simulation ended")
+}
+
+// TestMultiChequeSimulation just launches two nodes, a creditor and a debitor.
+// The debitor is the one owing to the creditor, so the debitor is the one sending cheques
+// It sends multiple cheques in a sequence to the same node.
+// The test checks that accounting still works properly afterwards and that
+// cheque cumulation values add up correctly
+func TestMultiChequeSimulation(t *testing.T) {
+	nodeCount := 2
+	// create the shared backend and params
+	params, err := newSharedBackendSwaps(nodeCount)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// cleanup backend
+	defer params.backend.Close()
+
+	// setup the wait for mined transaction function for testing
+	cleanup := setupContractTest()
+	defer cleanup()
+
+	params.backend.cashDone = make(chan struct{}, 1)
+	defer close(params.backend.cashDone)
+	// initialize the simulation
+	sim := simulation.NewBzzInProc(newSimServiceMap(params), false)
+	defer sim.Close()
+
+	log.Info("Initializing")
+
+	ctx, cancelSimRun := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancelSimRun()
+
+	_, err = sim.AddNodesAndConnectFull(nodeCount)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// define the nodes
+	debitor := sim.UpNodeIDs()[0]
+	creditor := sim.UpNodeIDs()[1]
+	// get the testService for the debitor
+	item := sim.MustNodeItem(debitor, bucketKeySwap)
+	debitorSvc := item.(*testService)
+
+	// get the testService for the creditor
+	peerItem := sim.MustNodeItem(creditor, bucketKeySwap)
+	creditorSvc := peerItem.(*testService)
+
+	for {
+		// let's always be nice and allow a time out to be catched
+		select {
+		case <-ctx.Done():
+			t.Fatal("Timed out waiting for all swap peer connections to be established")
+		default:
+		}
+		// the node has all other peers in its peer list
+		if len(debitorSvc.swap.peers) == 1 && len(creditorSvc.swap.peers) == 1 {
```

Comment:
Protect swap.peers for both debitorSvc and creditorSvc with swap.peersLock to avoid data races.